### DISCUSSION TITLE
How to use Quectel RM510 as UE

### DISCUSSION THREAD

**Discussion by initChat** [srsran/srsRAN_Project#426] [General Help] [answered]
I used to use UE as samsung a7 and successful with srsRAN_4G. And run srsRAN_Project with ZMQ as bladeRF not work very well with 5G.

Right now, for srsRAN_4G, I want to use Quectel RM510Q-GL with sysmocom sim card as UE.
Currently I can only ping using 'AT+QPING=1,"8.8.8.8" on minicom
when I run 'AT+QPING=1,"8.8.8.8"' on minicom
the screen show
```
at+qping=1,"8.8.8.8"                                                            
OK                                                                              
 +QPING: 569   
```
and the tracer of gNB (BladeRF) showing
```
               -----------------DL----------------|-------------------------UL-------------------------
rat  pci rnti  cqi  ri  mcs  brate   ok  nok  (%) | pusch  pucch  phr  mcs  brate   ok  nok  (%)    bsr
lte    1   62   15   1    5   2.3k    6    0   0% |  23.4   23.1   40   18    79k   11    0   0%    0.0
lte    1   62   15   1    0      0    0    0   0% |   n/a   99.9    0    0      0    0    0   0%    0.0
lte    1   62   15   1    0      0    0    0   0% |   n/a   99.9    0    0      0    0    0   0%    0.0
lte    1   62   15   1    0      0    0    0   0% |   n/a   99.9    0    0      0    0    0   0%    0.0
```
The bit rate only occur once and never occur again. But I woud like to test for longer time. Is there a command for that? for ping continuously?

However, if you guys don't use QPING command, what other command you can use, to make quectel run continuously on srsRAN?

Currently I use
SDR: BladeRF A4
UE: Quectel RM510Q-GL with sysmocom sim card
And run srsRAN_4G with a core network from open5GS

**Comment by initChat** [comment]
this is log of the gNB and quectel minicom
[gnb_log.txt](https://github.com/srsran/srsRAN_Project/files/14036413/gnb_log.txt)
[quectel_minicom.txt](https://github.com/srsran/srsRAN_Project/files/14036424/quectel_minicom.txt)

**Comment by pgawlowicz** [comment]
While the manual for Quectel RM510Q-GL says nothing about QPING, there is this manual for EC2x&EG9x&EM05T with QPING section:
https://sixfab.com/wp-content/uploads/2018/09/Quectel_EC2xEG9xEM05_TCPIP_AT_Commands_Manual_V1.0.pdf

It seems it is possible to specify number of pings:
![image](https://github.com/srsran/srsRAN_Project/assets/13730642/a6bfe0e1-6a3a-439a-9ce4-b93e32e16497)

Could you try the following command?
```
at+qping=1,"8.8.8.8",4,10
```

**Reply by herlesupreeth** [reply, marked as answer]
I use mmcli and use ping command from terminal. For example

First I run mmcli to list all connected modem

```
sudo mmcli -L
    /org/freedesktop/ModemManager1/Modem/0 [Intel] MBIM
```

then, take note of modem number i.e. in my case its 0. Then, I run following commands

```
sudo mmcli -m 0 -e
sudo mmcli -m 0 --simple-connect="apn=internet"
```

the above command initiates a PDN connection for DNN "internet". Then, you should be able to use AT command or see in logs of 5GC for the allocated IP for internet DNN. Say for example, IP allocated is 192.168.100.2, then check the output of `ip a ` and you should find an interface named `wwanX` e.g. wwan0. Assign the allocated IP to that interface as follows

```
sudo ip link set wwan0 up
sudo ip a add 192.168.100.2/24 dev wwan0
```

Now, you can ping as follows

`ping 8.8.8.8 -I wwan0`

**Reply by initChat** [reply]
@pgawlowicz this one does not work, it still produces 1 ping for me.

**Reply by initChat** [reply]
@herlesupreeth This one is working!!! Thank you very much.

**Reply by chonghua1** [reply]
Dear @herlesupreeth,
I tried this method with Quectel RM500Q-GL, when I ping as `ping 8.8.8.8 -I wwan0` I can't get response, but I can see there is some traffic at the gNB console:
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/b1d23311-ebca-4aba-b8c0-66eb1a7b49f7)
and Open5GS core network logs some error:
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/7f793200-1c2f-45e3-af61-7824ea9a4684)
Do you have any idea of this issue? Thank you!

**Reply by herlesupreeth** [reply]
Then, please verify you have NATing enabled on the 5GC. If then it does not work then send me the output of 5GC (like above screenshot) and output of ip a on the machine to which Quectel modem is connected.

Can you also describe your setup? Are you running gNB + 5GC using the docker compose present in this repo?

**Reply by initChat** [reply]
@chonghua1 Can I ask the result of your `ifconfig wwan0`. Sometimes I didn't set ip correctly.
Did you skip step to assign ip?
```
sudo ip link set wwan0 up
sudo udhcpc -q -f -i wwan0
```
Also, you can use wireshark on the computer that run gNB as well. Sometimes traffic that is shown on gNB trace is retransmission and try to ping, but cannot ping successfully.

**Reply by chonghua1** [reply]
> Then, please verify you have NATing enabled on the 5GC. If then it does not work then send me the output of 5GC (like above screenshot) and output of ip a on the machine to which Quectel modem is connected.
> 
> Can you also describe your setup? Are you running gNB + 5GC using the docker compose present in this repo?

I use open5GS docker as 5GC, and the corenetwork is deployed with gNB at the same machine A, the Quectel module connect to another machine B.
I have set the NAT rule at 5GC host use the command:
```
sudo sysctl -w net.ipv4.ip_forward=1
sudo iptables -t nat -A POSTROUTING -o <IFNAME> -j MASQUERADE
```
the route table at machine A (5GC host) like this :
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/d13dc573-547d-4b9c-8334-04f31e5c16fc)

the ` ip a ` output at machine B (UE host) like this:
![default](https://github.com/srsran/srsRAN_Project/assets/42130689/47000011-b880-4af9-9489-34201e9b5b0c)

when I use AT commands at UE host like this:
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=8239050822025841500 skey=@crypt_777de82d_650381097698cb6d751d3d9dd474b323 mmweb_appid=wx_webfilehelper](https://github.com/srsran/srsRAN_Project/assets/42130689/1f0e798e-6616-487c-9b34-39ead93f849b)

the output of 5GC is:
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/04c26011-5dc0-4d11-aff1-3d32a58fbad2)

but when use the mmcli and use ping command from terminal, the 5GC output is still with some UPF errors:
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/bfa0e2ca-e428-4fe0-86ce-69070c722a85)

**Reply by chonghua1** [reply]
> @chonghua1 Can I ask the result of your `ifconfig wwan0`. Sometimes I didn't set ip correctly. Did you skip step to assign ip?
> 
> ```
> sudo ip link set wwan0 up
> sudo udhcpc -q -f -i wwan0
> ```
> 
> Also, you can use wireshark on the computer that run gNB as well. Sometimes traffic that is shown on gNB trace is retransmission and try to ping, but cannot ping successfully.

Hi @pchat-imm,
Yes ,I have allocated the ip address for wwan0 using command
```
sudo ip link set wwan0 up
sudo ip a add 192.168.40.2/24 dev wwan0
```
and when I use QPING from AT command, the wireshark output is:
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/e06818c9-513a-4b1e-8077-e4be445780d5)

when I use mmcli and use ping command from terminal, the wireshark output is:
![image](https://github.com/srsran/srsRAN_Project/assets/42130689/2d03cd10-4bbe-4c28-85f9-4acb62a400bd)
and the 5GC logs some UPF Errors

**Reply by chonghua1** [reply]
> @chonghua1 Can I ask the result of your `ifconfig wwan0`. Sometimes I didn't set ip correctly. Did you skip step to assign ip?
> 
> ```
> sudo ip link set wwan0 up
> sudo udhcpc -q -f -i wwan0
> ```
> 
> Also, you can use wireshark on the computer that run gNB as well. Sometimes traffic that is shown on gNB trace is retransmission and try to ping, but cannot ping successfully.

the output of` ifconfig wwan0` at UE host like this:
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=9091177402573410435 skey=@crypt_777de82d_b43c6d9a75a31069103a194f481efc00 mmweb_appid=wx_webfilehelper](https://github.com/srsran/srsRAN_Project/assets/42130689/5f094ba3-0a53-4275-8c58-befd9984f94e)

**Reply by herlesupreeth** [reply]
Wait, i am confused. Why are you assigning 192.168.40.2 to wwan0 while your UE got the IP 10.45.1.6 from the 5GC. You should be assigning 10.45.1.6 to wwan0

**Reply by chonghua1** [reply]
> Wait, i am confused. Why are you assigning 192.168.40.2 to wwan0 while your UE got the IP 10.45.1.6 from the 5GC. You should be assigning 10.45.1.6 to wwan0

Ah I thought it was given by random, how dumb I am... Now it's worked, thank you! :)
